pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1048 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/agent/util"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+	"antrea.io/antrea/pkg/util/channel"
+	"antrea.io/antrea/pkg/util/k8s"
+	utilsets "antrea.io/antrea/pkg/util/sets"
+)
+
+const (
+	controllerName = "TrafficControlController"
+	// How long to wait before retrying the processing of a TrafficControl change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing a TrafficControl change.
+	defaultWorkers = 4
+	// Disable resyncing.
+	resyncPeriod time.Duration = 0
+
+	// Default VXLAN tunnel destination port.
+	defaultVXLANTunnelDestinationPort = int32(4789)
+	// Default Geneve tunnel destination port.
+	defaultGENEVETunnelDestinationPort = int32(6081)
+
+	portNamePrefixVXLAN  = "vxlan"
+	portNamePrefixGENEVE = "geneve"
+	portNamePrefixGRE    = "gre"
+	portNamePrefixERSPAN = "erspan"
+
+	trafficControlTunnelType = "tc-tunnel-type"
+
+	maxRetryForHostLink = 5
+)
+
+// trafficControlState keeps the actual state of a TrafficControl that has been realized.
+type trafficControlState struct {
+	// The actual target port of a TrafficControl.
+	targetPort uint32
+	// The actual return port of a TrafficControl.
+	returnPort uint32
+	// The actual action of a TrafficControl.
+	action v1alpha2.TrafficControlAction
+	// The actual direction of a TrafficControl.
+	direction v1alpha2.Direction
+	// The actual openflow ports for which we have installed TrafficControl rules.
+	ofPorts sets.Int32
+	// The actual Pods of the TrafficControl. Used to identify stale Pods when updating or deleting a TrafficControl.
+	pods sets.String
+}
+
+type Controller struct {
+	ofClient openflow.Client
+
+	ovsBridgeClient    ovsconfig.OVSBridgeClient
+	ovsPortUpdateMutex sync.Mutex
+
+	interfaceStore interfacestore.InterfaceStore
+
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	namespaceInformer     cache.SharedIndexInformer
+	namespaceLister       corelisters.NamespaceLister
+	namespaceListerSynced cache.InformerSynced
+
+	tcStates      map[string]*trafficControlState
+	tcStatesMutex sync.RWMutex
+
+	podToTCBindings      map[string]*podToTCBinding
+	podToTCBindingsMutex sync.RWMutex
+
+	portToTCBindings      map[uint32]*portToTCBinding
```

Comment:
This seems a secondary index of `tcStates`. Maybe change `tcStates` to cache.Indexer with two indices: returnPort and targetPort to avoid maintaining two storages and potential inconsistence.

Reply:
any comment about this one? With it, the two extra mapping are not requred.

Reply:
Updated as you said.